### DISCUSSION TITLE
Optimizing Drawing

### DISCUSSION THREAD

**Discussion by Ruckamongus** [switchboy/IsoRTS#22] [General]
Considerable work has been made on the bottlenecks of screen redrawing by minimizing draw() calls, but if it becomes a problem we can minimize draw calls via batching. This would require:

- Texture packing: placing multiple spirites on a small number of textures (ideally 1) to avoid texture swaps (loading and unloading the active texture from RAM to GPU RAM
- Vertex array: "draw" sprites to a vertex array and draw the vertex array. This can dramatically decrease draw calls

I've done this before and am willing to do this if needed, but I don't think there's enough draw() calls yet for this amount of work to be worthwhile.

**Comment by switchboy** [comment]
Once prototyping is done and there is a definitive selection of buildings objects and actors I was planning on combining all te sprites into one big texture file for each category. For now I want te leave it as is since it allows for rapid development.

At te moment i am working on object, building and actor templates. This will allow the addition of new content by just editing one file. And adding the button for creation.

**Comment by Ruckamongus** [comment]
Round two of running the profiler. It seems that switching between threads is the hottest function followed by pathfinding. Just an FYI:
![image](https://user-images.githubusercontent.com/1320779/129449623-fc793727-4e53-4745-a9e7-e04eaaacf9b5.png)

**Reply by switchboy** [reply]
Maybe the concurrent process creation and destruction is creating more overhead then the performance that is gained by it is worth?

 I needed it in a very early prototype which had a very inefficient A* algorithm. After revision de algorithm is much faster.

**Comment by Ruckamongus** [comment]
> Maybe the concurrent process creation and destruction is creating more overhead then the performance that is gained by it is worth?
I think this is the case. Maybe try a branch with and without it. Then we can run the profiler and see.

**Comment by switchboy** [comment]
I made a branch without the async didn't notice performance degradation in fact there was less stuttering. What I notice that sorting the list of cells (to get the one with the lowest estimated cost to goal) is taking a lot CPU time.  (it sorts std::list<Cells*> listToCheck;)

So going of of this it might make sense to try and see if a vector makes more sense. Since it is only a list of pointers the 'data' portion of the list is small so going off those benchmarks sorting a vector which contains only pointers should be **much** faster
